### PR TITLE
Containerservices addonprofile

### PR DESCRIPTION
**Pull request by lfshr**
Added addon_profile functionality to kubernetes clusters

### CODE REVIEW

File: azurerm/resource_arm_kubernetes_cluster.go [hunk 1]
```diff
@@ -566,6 +626,35 @@ func expandAzureRmKubernetesClusterAgentProfiles(d *schema.ResourceData) []conta
...
+
+		name := config["name"].(string)
+		enabled := config["enabled"].(bool)
+
+		// We don't know the keys explicitly, they could be anything.
+		addonConfig := config["config"].(map[string]interface{})
+		configValue := make(map[string]*string)
+		for k, v := range addonConfig {
+			configEntry := v.(string)
+			configValue[k] = &configEntry
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_kubernetes_cluster.go]: _minor_ we can combine these to make: `utils.String(v.(string))`

File: azurerm/resource_arm_kubernetes_cluster.go [hunk 2]
```diff
@@ -209,6 +209,33 @@ func resourceArmKubernetesCluster() *schema.Resource {
 				Set: resourceAzureRMKubernetesClusterServicePrincipalProfileHash,
 			},
 
+			"addon_profile": {
```
Comment (tombuildsstuff) [inline azurerm/resource_arm_kubernetes_cluster.go]: can we document these new fields in the docs? this file's located at `./website/docs/r/kubernetes_cluster.html`

Comment (tombuildsstuff) [reply azurerm/resource_arm_kubernetes_cluster.go]: can we also add an acceptance test covering these new fields?

File: azurerm/resource_arm_kubernetes_cluster.go [hunk 3]
```diff
@@ -469,6 +505,30 @@ func flattenAzureRmKubernetesClusterAccessProfile(profile *containerservice.Mana
 	return nil, []interface{}{}
 }
 
+func flattenAzureRmKubernetesClusterAddonProfiles(profile map[string]*containerservice.ManagedClusterAddonProfile) []interface{} {
+	values := make([]interface{}, 0)
+	for k, v := range profile {
+		addonProfile := map[string]interface{}{
+			"name":    k,
+			"enabled": *v.Enabled,
```
[the remark this answers is not in the document]
Comment (lfshr) [reply azurerm/resource_arm_kubernetes_cluster.go]: Is that still the case on a Required field, should I do some kind of error handling here if it's nil or is the nil check all that's needed? Good spot. I totally missed that.

[the remark this answers is not in the document]
Comment (tombuildsstuff) [reply azurerm/resource_arm_kubernetes_cluster.go]: this is the response coming back from the API, which can be nil due to bugs in the API / changes in the schema over time (this is something we've learnt over time, unfortunately) - so we need to check it either way unfortunately

**Comment by lfshr** [comment]
@tombuildsstuff Code updated to reflect your comments. This PR should be good to go.

**Comment by lfshr** [comment]
@tombuildsstuff now that advanced networking is done I'm turning my attention to this. Can you have a look at the latest commit (cea7faa) and tell me if that's more what you're looking for? Only have HttpApplicationRouting enabled at the moment.

I'm going on holiday on Saturday for two weeks. I'll try my best to finish this but if I fail, feel free to continue.

**Comment by tombuildsstuff** [comment]
hey @lfshr 

Sorry for the delayed re-review here - I've taken a look into this and merged master into this branch / have resolved the merge conflicts by adding the changes from this PR back in. Since I don't have permission to push to your fork - I've had to push to this repo and have opened #1751 which includes your original commits plus the updates required to get this merged - I hope you don't mind!

Thanks!

**Comment by lfshr** [comment]
@tombuildsstuff no problem at all. Thanks very much for continuing this. I ran out of time before my holiday I'm afraid :disappointed:

Arrived back last night so let me know if there's anything else you need me to do.

**Comment by tombuildsstuff** [comment]
@lfshr all good - apologies for the delay getting this merged; we've been heads-down trying to work through some larger cross-provider items (e.g. #1746 / #1728) over the past couple of weeks

**Pull request closed without merging**